### ISSUE TITLE
FIM (fill in the middle) support for inline completer on typing

### ISSUE THREAD

**Issue by krassowski** [jupyterlab/jupyterlab#17060]
### Problem

_Fill-in-the-middle_ allows to add code in the middle of the code line. This is currently supported when manually invoking the inline completer, but it is not possible for user to get inline completion suggestions automatically when typing in the middle of the line (these are only returned at the end of the line); this is codified in:

https://github.com/jupyterlab/jupyterlab/blob/bb596b37d44cacb7aaec2c09aa5d6c8b43a3a2a8/packages/completer/src/handler.ts#L467-L473

### Proposed Solution

Add an opt-in option allowing for inline completion to be auto-triggered when anywhere in the line.

Of note, the manual invoke is allowed anywhere since https://github.com/jupyterlab/jupyterlab/pull/16298 included `4.1.x`.

### Additional context

Quoting the Codeium [blog post](https://codeium.com/blog/inline-fim-code-suggestions) (emphasis mine):

> This scenario where there is text on both sides of the cursor (on the same line) and we want autocomplete suggestions is a special case of the fill-in-the-middle task called in-line fill-in-the-middle, or in-line FIM for short. **And it is not easy to get in-line FIM to work well enough to be helpful rather than distracting.**
> 
> [...]
> So what is likely going on with these AI code assistants? In the original post, we showed how TabNine didn’t have generic FIM capabilities in the first place. Amazon CodeWhisperer on face value does, but they must have made some significant fundamental errors in building FIM since the suggestions are quite poor, error-prone, and often not referencing post-cursor context.
> 
> But what about Github Copilot? They have reasonable FIM suggestions for filling in lines and using future context, so why is in-line FIM not enabled? Honestly, we are not sure. Copilot has normal FIM suggestions and it seems like there shouldn’t be anything that is stopping it from exposing in-line FIM suggestions. Our guess is that they tried to enable in-line FIM, but the performance wasn’t good enough, maybe from a lack of in-line FIM-like training examples or some other training knobs. All we know is that for now, Copilot has not enabled in-line FIM in production, and likely that means there is something up with the underlying model, which they are unwilling to retrain, or their FIM handling capabilities are inherently worse.

This appears like a good enough rationale for making it opt-in in JupyterLab rather than opt-out, especially because compared to GitHub Copilot or Codeium, JupyterLab does not have a set of curated supported models for which pre-processing and context provision is fine-tuned to return good enough in-line FIM to be more useful than distracting.

---

Previously a broader question about FIM was opened in https://github.com/jupyterlab/jupyterlab/issues/16671 but it did not contain specific change requests - only a question about support status.

--- 

LSP specification does not have a special trigger kind or metadata for in-line FIM.

**Comment by Darshan808** [comment]
Hi Mike,

I’m trying to understand this issue but am facing some confusion. When I tested the autocompleter, it suggested completions even in the middle of a line (when the word is separated from others by spaces). From the code, I noticed this happens due to the _onTextChange event, which calls _makeRequest() to display completions.

![image](https://github.com/user-attachments/assets/5eb0326b-fea5-4ebf-991e-6b355d6fb5df)

![image](https://github.com/user-attachments/assets/72002b4f-bed6-4a3a-901e-944463096a5c)

Could you please provide an example snippet where the autocompleter does not show completions in the middle of a line automatically but does so when triggered manually (e.g., with the Tab key)? This will help me better understand the problem and work toward a solution.

Thank you for your guidance!

**Comment by krassowski** [comment]
This issue is not about the auto-completer, but inline completer. You need to enable inline completer via settings (it is opt-in). You can get AI-generated inline completions by installing `jupyter-ai` and enabling it (requires configuring a model).

**Comment by Darshan808** [comment]
Understood. Thank you for the clarification!

**Comment by krassowski** [comment]
The heavy lift in this issue will be adopting our history inline completions provider so that we can test it in core (without depending on AI models which are not great for testing as these generally return different result each time*/are expensive to run)

* well.... this can be overcome for some models by setting temperature to zero.

**Comment by MMesch** [comment]
The triaging agent has identified this issue as likely completed. I just tested and after activation of the fill in the middle option in settings, I can indeed do this:

<img width="376" height="127" alt="Image" src="https://github.com/user-attachments/assets/6a657ced-2874-47f8-98a9-abee6ccedcbe" />

It therefore seems that this feature is indeed done. Could you close if that's correct @krassowski or otherwise leave a comment making explicit what is still missing?

---

**Agent Rationale**: The requested feature has been successfully implemented and merged in PR #17067. The `autoFillInMiddle` setting is now available for inline completion providers, allowing them to opt-in to providing completions when the cursor is in the middle of a line during automatic triggering. This directly addresses the enhancement request made in the issue.

**Comment by krassowski** [comment]
Yeah. We should probably do document it too but let's close this issue.
